File: backend/src/main/java/com/devrats/service/AuthService.java
package com.devrats.service;

import com.devrats.model.RefreshToken;
import com.devrats.model.User;
import com.devrats.model.Squad;
import com.devrats.repository.RefreshTokenRepository;
import com.devrats.repository.UserRepository;
import com.devrats.repository.ScoreRepository;
import com.devrats.repository.NotificationRepository;
import com.devrats.repository.SquadMemberRepository;
import com.devrats.repository.SquadRepository;
import com.devrats.security.JwtProvider;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.UUID;

@Service
public class AuthService {
    private final UserRepository userRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final JwtProvider jwtProvider;
    private final LeagueService leagueService;
    private final ScoreRepository scoreRepository;
    private final NotificationRepository notificationRepository;
    private final SquadMemberRepository squadMemberRepository;
    private final SquadRepository squadRepository;

    public AuthService(UserRepository userRepository, RefreshTokenRepository refreshTokenRepository, JwtProvider jwtProvider, LeagueService leagueService, ScoreRepository scoreRepository, NotificationRepository notificationRepository, SquadMemberRepository squadMemberRepository, SquadRepository squadRepository) {
        this.userRepository = userRepository;
        this.refreshTokenRepository = refreshTokenRepository;
        this.jwtProvider = jwtProvider;
        this.leagueService = leagueService;
        this.scoreRepository = scoreRepository;
        this.notificationRepository = notificationRepository;
        this.squadMemberRepository = squadMemberRepository;
        this.squadRepository = squadRepository;
    }

    @Transactional
    public void deleteAccount(String userId) {
        User user = userRepository.findById(userId).orElseThrow(() -> new RuntimeException("User not found"));
        
        // 1. Soft delete the user
        user.setDeletedAt(Instant.now());
        userRepository.save(user);

        // 2. Delete refresh tokens to force immediate logout across all devices
        refreshTokenRepository.deleteByUserId(userId);
    }

    @Transactional
    public AuthResponse findOrCreateUser(String githubId, String username, String displayName, String avatarUrl, String email) {
        User user = userRepository.findByGithubId(githubId);
        if (user == null) {
            user = new User();
            user.setId(UUID.randomUUID().toString());
            user.setGithubId(githubId);
            user.setUsername(username);
            user.setDisplayName(displayName);
            user.setAvatarUrl(avatarUrl);
            user.setEmail(email);
            user = userRepository.save(user);
        } else if (user.getDeletedAt() != null) {
            // Restore soft-deleted account
            user.setDeletedAt(null);
            user = userRepository.save(user);
        }

        leagueService.assignUserToGroup(user);

        String accessToken = jwtProvider.generateAccessToken(user.getId());
        String refreshToken = jwtProvider.generateRefreshToken(user.getId());

        RefreshToken rt = new RefreshToken();
        rt.setId(UUID.randomUUID().toString());
        rt.setUserId(user.getId());
        rt.setToken(refreshToken);
        rt.setExpiresAt(Instant.now().plus(7, ChronoUnit.DAYS));
        refreshTokenRepository.save(rt);

        return new AuthResponse(
                accessToken,
                refreshToken,
                new UserProfile(
                        user.getId(),
                        user.getUsername(),
                        user.getDisplayName(),
                        user.getAvatarUrl(),
                        user.getTotalScore(),
                        user.getEffectiveStreak(),
                        user.getBestStreak(),
                        user.getLeague()
                )
        );
    }

    @Transactional
    public TokenRefreshResponse refreshAccessToken(String token) {
        RefreshToken stored = refreshTokenRepository.findByToken(token);
        if (stored == null || stored.getExpiresAt().isBefore(Instant.now())) {
            return null;
        }
        String userId = stored.getUserId();
        
        // Check if user is soft-deleted
        User user = userRepository.findById(userId).orElse(null);
        if (user == null || user.getDeletedAt() != null) {
            refreshTokenRepository.delete(stored);
            return null;
        }
        
        refreshTokenRepository.delete(stored);

        String newAccess = jwtProvider.generateAccessToken(userId);
        String newRefresh = jwtProvider.generateRefreshToken(userId);

        RefreshToken rt = new RefreshToken();
        rt.setId(UUID.randomUUID().toString());
        rt.setUserId(userId);
        rt.setToken(newRefresh);
        rt.setExpiresAt(Instant.now().plus(7, ChronoUnit.DAYS));
        refreshTokenRepository.save(rt);

        return new TokenRefreshResponse(newAccess, newRefresh);
    }

    public UserProfile getProfile(String userId) {
        User user = userRepository.findById(userId).orElse(null);
        if (user == null || user.getDeletedAt() != null) {
            return null;
        }
        return new UserProfile(
                user.getId(),
                user.getUsername(),
                user.getDisplayName(),
                user.getAvatarUrl(),
                user.getTotalScore(),
                user.getEffectiveStreak(),
                user.getBestStreak(),
                user.getLeague()
        );
    }

    public record AuthResponse(String accessToken, String refreshToken, UserProfile user) {}
    public record TokenRefreshResponse(String accessToken, String refreshToken) {}
    public record UserProfile(String id, String username, String displayName, String avatarUrl, Integer totalScore, Integer currentStreak, Integer bestStreak, String league) {}
}

File: backend/src/main/kotlin/com/devrats/features/auth/AuthService.kt
package com.devrats.features.auth

import com.devrats.features.auth.models.Users
import com.devrats.features.auth.models.AuthResponse
import com.devrats.features.auth.models.UserProfile
import com.devrats.features.auth.models.TokenRefreshResponse
import com.devrats.shared.security.JwtProvider
import com.devrats.shared.exceptions.UnauthorizedException
import kotlinx.datetime.Clock
import kotlin.time.Duration.Companion.milliseconds

class AuthService(private val repo: AuthRepository, private val jwt: JwtProvider) {

    fun findOrCreateUser(githubId: String, username: String, displayName: String, avatarUrl: String?): AuthResponse {
        val existing = repo.findByGithubId(githubId)
        val userId = if (existing != null) existing[Users.id] else repo.createUser(githubId, username, displayName, avatarUrl)
        val accessToken = jwt.generateAccessToken(userId)
        val refreshToken = jwt.generateRefreshToken(userId)
        val expiresAt = Clock.System.now().plus(jwt.refreshExp.milliseconds)
        repo.saveRefreshToken(userId, refreshToken, expiresAt)
        val user = repo.findById(userId)!!
        return AuthResponse(
            accessToken = accessToken, 
            refreshToken = refreshToken,
            user = UserProfile(
                id = user[Users.id], 
                username = user[Users.username],
                displayName = user[Users.displayName], 
                avatarUrl = user[Users.avatarUrl],
                totalScore = user[Users.totalScore], 
                currentStreak = user[Users.currentStreak], 
                bestStreak = user[Users.bestStreak]
            )
        )
    }

    fun refreshAccessToken(refreshToken: String): TokenRefreshResponse {
        val stored = repo.findRefreshToken(refreshToken) ?: throw UnauthorizedException("Invalid refresh token")
        val userId = stored[com.devrats.features.auth.models.RefreshTokens.userId]
        repo.deleteRefreshToken(refreshToken)
        val newAccess = jwt.generateAccessToken(userId)
        val newRefresh = jwt.generateRefreshToken(userId)
        val expiresAt = Clock.System.now().plus(jwt.refreshExp.milliseconds)
        repo.saveRefreshToken(userId, newRefresh, expiresAt)
        return TokenRefreshResponse(accessToken = newAccess, refreshToken = newRefresh)
    }

    fun getProfile(userId: String): UserProfile {
        val user = repo.findById(userId) ?: throw UnauthorizedException("User not found")
        return UserProfile(
            id = user[Users.id], 
            username = user[Users.username],
            displayName = user[Users.displayName], 
            avatarUrl = user[Users.avatarUrl],
            totalScore = user[Users.totalScore], 
            currentStreak = user[Users.currentStreak], 
            bestStreak = user[Users.bestStreak]
        )
    }
}
